src/volume_mgr.rs

```diff
@@ -878,7 +878,9 @@ where
                 Err(e) => return Err(e),
             };
             let to_copy = core::cmp::min(block_avail, bytes_to_write - written);
-            let block = if block_offset != 0 || data.open_files[file_idx].current_offset < data.open_files[file_idx].entry.size {
+            let block = if block_offset != 0
```

Comment:
I think I would prefer something like:

```rust
            let block = if (block_offset == 0) && (to_copy == Block::LEN) {
                // we're replacing the whole Block, so the previous contents
                // are irrelevant
                data.block_cache.blank_mut(block_idx)
            } else {
                debug!("Reading for partial block write");
                data.block_cache
                    .read_mut(block_idx)
                    .map_err(Error::DeviceError)?
            };
```

This also passes your new test case.

Reply:
I believe that your fix would mean a write of 512 bytes at the start of the file would do a read-modify-write, when a write is sufficient.

Reply:
Will this work also at EOF write happens to be exactly on block boundary and less than Block::LEN?
In that case won't there be in the block_cache potentially data from previous reads? I thought that was the reason for filling blank in the first place, to support appending and clearing data if EOF was right at the beginning of a new block.

Reply:
So actually, I could be persuaded to make this `if (block_offset == 0) && (to_copy == block_avail)`.

Reply:
So I replied to your erased comment. Deleting my replies.

Reply:
I don't have any preference, not really very familiar with the code. Let me know your preference and I'll modify it, I trust your judgement better than mine here.

Reply:
`to_copy` is the amount of stuff we are writing to this block (see the slice copy that happens after). `block_avail` is how much stuff in this block (starting at the beginning of the block, because you cannot have holes in files when using FAT) is actually valid.

This might be better:

```rust
            let block = if (block_offset == 0) && (to_copy == block_avail) {
                // we're replacing the whole Block, so the previous contents
                // are irrelevant
                data.block_cache.blank_mut(block_idx)
            } else {
                debug!("Reading for partial block write");
                data.block_cache
                    .read_mut(block_idx)
                    .map_err(Error::DeviceError)?
            };
```

Reply:
Consider a file that is 1026 bytes long. We write 514 bytes starting at offset 511.

We should:

1. Read-modify-write Block 0 in the file, replacing `block[511]`
2. Write Block 1 in the file.
3. Read-modify write Block 2 in the file, replacing `block[0]` but leaving `block[1]` unchanged (and leaving `block[2..512]` unchanged as well).

Consider a file that is 1026 bytes long. We write 515 bytes starting at offset 511.

We should:

1. Read-modify-write Block 0 in the file, replacing `block[511]`
2. Write Block 1 in the file
3. Write Block 2 - it contains two data bytes at `block[0..2]` and 510 zero bytes after that, but the zeroes are beyond the EOF.

Reply:
I made the update based on the above and also updated the changelog. 
Anything else required?